metadata-ingestion/src/datahub/ingestion/api/source_helpers.py

```diff
@@ -366,3 +370,85 @@ def _prepend_platform_instance(
         return [BrowsePathEntryClass(id=urn, urn=urn)] + entries
 
     return entries
+
+
+def auto_incremental_lineage(
+    graph: Optional[DataHubGraph],
+    incremental_lineage: bool,
+    include_column_level_lineage: bool,
+    stream: Iterable[MetadataWorkUnit],
+) -> Iterable[MetadataWorkUnit]:
+    if not incremental_lineage:
+        yield from stream
+        return  # early exit
+
+    for wu in stream:
+        lineage_aspect: Optional[UpstreamLineageClass] = wu.get_aspect_of_type(
+            UpstreamLineageClass
+        )
+        urn = wu.get_urn()
+        # TODO: use same systemMetadata as before
+
+        if lineage_aspect:
+            if isinstance(wu.metadata, MetadataChangeEventClass):
+                set_aspect(
+                    wu.metadata, None, UpstreamLineageClass
+                )  # we'll emit upstreamLineage separately below
+                if len(wu.metadata.proposedSnapshot.aspects) > 0:
+                    yield wu
+
+            yield _lineage_wu_via_read_modify_write(
+                graph, urn, lineage_aspect
+            ) if include_column_level_lineage else _patch_lineage_wu_for_table_level_lineage(
+                urn, lineage_aspect
+            )
+        else:
+            yield wu
+
+
+def _patch_lineage_wu_for_table_level_lineage(
```

Comment:
might even be worth adding this as a static helper method e.g. `DatasetPatchBuilder.from_table_upstream_lineage(...)`

```suggestion
def _convert_upstream_lineage_to_patch(
```